bake/hcl.go

```diff
@@ -171,79 +185,325 @@ func (sc *StaticConfig) Values(withEnv bool) (map[string]cty.Value, error) {
 		}
 	}
 
-	sc.values = map[string]cty.Value{}
-	sc.progress = map[string]struct{}{}
+	sc.funcDefs = map[string]*Function{}
+	for _, v := range sc.Functions {
+		sc.funcDefs[v.Name] = v
+	}
 
+	sc.ectx = hcl.EvalContext{
+		Variables: map[string]cty.Value{},
+		Functions: stdlibFunctions,
+	}
+	sc.funcs = map[string]function.Function{}
+	sc.progress = map[string]struct{}{}
+	sc.progressF = map[string]struct{}{}
 	for k := range sc.attrs {
-		if _, err := sc.resolveValue(k); err != nil {
+		if err := sc.resolveValue(k); err != nil {
 			return nil, err
 		}
 	}
 
 	for k := range sc.defaults {
-		if _, err := sc.resolveValue(k); err != nil {
+		if err := sc.resolveValue(k); err != nil {
 			return nil, err
 		}
 	}
-	return sc.values, nil
+
+	for k := range sc.funcDefs {
+		if err := sc.resolveFunction(k); err != nil {
+			return nil, err
+		}
+	}
+	return &sc.ectx, nil
 }
 
-func (sc *StaticConfig) resolveValue(name string) (v *cty.Value, err error) {
-	if v, ok := sc.values[name]; ok {
-		return &v, nil
+type jsonExp interface {
+	ExprList() []hcl.Expression
+	ExprMap() []hcl.KeyValuePair
+}
+
+func elementExpressions(je jsonExp, exp hcl.Expression) []hcl.Expression {
+	list := je.ExprList()
+	if len(list) != 0 {
+		exp := make([]hcl.Expression, 0, len(list))
+		for _, e := range list {
+			if je, ok := e.(jsonExp); ok {
+				exp = append(exp, elementExpressions(je, e)...)
+			}
+		}
+		return exp
 	}
-	if _, ok := sc.progress[name]; ok {
-		return nil, errors.Errorf("variable cycle not allowed")
+	kvlist := je.ExprMap()
+	if len(kvlist) != 0 {
+		exp := make([]hcl.Expression, 0, len(kvlist)*2)
+		for _, p := range kvlist {
+			exp = append(exp, p.Key)
+			if je, ok := p.Value.(jsonExp); ok {
+				exp = append(exp, elementExpressions(je, p.Value)...)
+			}
+		}
+		return exp
 	}
-	sc.progress[name] = struct{}{}
+	return []hcl.Expression{exp}
+}
 
-	defer func() {
-		if v != nil {
-			sc.values[name] = *v
+func jsonFuncCallsRecursive(exp hcl.Expression) ([]string, error) {
+	je, ok := exp.(jsonExp)
+	if !ok {
+		return nil, errors.Errorf("invalid expression type %T", exp)
+	}
+	m := map[string]struct{}{}
+	for _, e := range elementExpressions(je, exp) {
+		if err := appendJSONFuncCalls(e, m); err != nil {
+			return nil, err
 		}
-	}()
+	}
+	arr := make([]string, 0, len(m))
+	for n := range m {
+		arr = append(arr, n)
+	}
+	return arr, nil
+}
 
-	def, ok := sc.attrs[name]
+func appendJSONFuncCalls(exp hcl.Expression, m map[string]struct{}) error {
+	v := reflect.ValueOf(exp)
```

Comment:
if you want to make this hackier but more readable, you might as well abuse //go:linkname to import those private structs.

Reply:
not that it is any cleaner but I don't think it would even work. I don't just need access to a private function/var in here.